### DISCUSSION TITLE
Support to PMU

### DISCUSSION THREAD

**Discussion by Everaldo-Gomes** [bao-project/bao-hypervisor#139] [General]
Hello,

Is there a way to access the PMU?

**Comment by danielRep** [comment]
Hi @Everaldo-Gomes ,

Could you be more explicit about what you are trying to do? Are you trying to access PMU from bao or the guest? What is your use-case?

**Reply by Everaldo-Gomes** [reply]
I'm running the Dual-guest Linux+FreeRTOS, using a Raspberry Pi 4.
I'd like to know the best way to access the PMU, is it from Bao or the guest?

**Reply by danielRep** [reply]
You should be able to access it from the guest. @Diogo21Costa could you clarify this?

**Reply by Diogo21Costa** [reply]
Hey @Everaldo-Gomes,

As @danielRep mentioned, you can access the PMU either from the guest or through the hypervisor. If you're planning to run some tests (benchmarks), it's better to access the PMU from the guest. But if you're thinking of setting up something at the hypervisor level, accessing the PMU is the way to go.

For example, with your current setup, you can update the PMU node in the Linux devicetree and then use the perf tool to track events from the PMU from the Linux OS guest. This requires two changes:

1. Include the PMU node in the Linux devicetree. For the rpi4 it would be something like this:
```dts
pmu {
        compatible = "arm,armv8-pmuv3";
        interrupt-parent = <0x4>;
        interrupts = <0 48 4>,
                 <0 49 4>,
                 <0 50 4>,
                 <0 51 4>;
};
```
2. Rebuild Linux to include the perf tool (follow [this instructions](https://gitlab.com/ESRGv3/shedding-light-static-partitioning-hypervisors/-/tree/master/guests/linux?ref_type=heads) to build perf and add it to your Linux image).


With these changes, you should be able to access the PMU from your Linux guest. However, if you need to access the PMU through FreeRTOS or a baremetal guest, or even at the hypervisor level, you can do so by using the `MSR` and `MRS` instructions to read and write to the following registers:

- PMCR_EL0
- PMXEVTYPER_EL0
- PMXEVCNTR_EL0
- PMCNTENSET_EL0
- PMCNTENCLR_EL0

Please let me know if you need further assistance!

**Reply by Everaldo-Gomes** [reply]
Thanks for your reply, I'll try it!
